File: src/engine/new-car.js
/**
 * Flow 2A (new car acquisition) model.
 *
 * Port of validation/new_car.py, sharing its locked conventions. The UI must
 * call this module rather than reimplementing any formula.
 */

import {
  annualCostFlows,
  compound,
  discountFactor,
  exponentialValue,
  levelMonthlyCost,
  loanBalance,
  loanPayment,
  monthlyRate,
  npv,
} from './finance.js';

export const RETURN = 'return';
export const BUYOUT = 'buyout';

/**
 * @typedef {Object} Scenario
 * @property {{ msrp: number, resaleAnchorMonth: number, resaleAnchorValue: number }} vehicle
 * @property {number} outTheDoor
 * @property {{ annualGovernmentCharges: number, annualEnergy: number }} ownership
 * @property {{ available: number, grossAnnualReturn: number }} capital
 * @property {number} discountRateAnnual
 * @property {{ downPayment: number, apr: number, termMonths: number }} [loan]
 * @property {Object} [lease]
 */

/**
 * Market value of the vehicle at a given month.
 * @param {Scenario['vehicle']} vehicle
 * @param {number} month
 * @returns {number}
 */
export function marketValue(vehicle, month) {
  return exponentialValue(
    vehicle.msrp,
    vehicle.resaleAnchorMonth,
    vehicle.resaleAnchorValue,
    month,
  );
}

function annualOwnershipTotal(ownership) {
  return ownership.annualGovernmentCharges + ownership.annualEnergy;
}

function notComparable(path, horizonMonths, reason) {
  return { path, horizonMonths, comparable: false, reason };
}

/**
 * Shared finalization: NPV, level monthly cost, overlay balances.
 */
function finalize(path, scenario, horizonMonths, flows, timeZeroOutlay, vehicleEquity) {
  const rate = monthlyRate(scenario.discountRateAnnual);
  const costNpv = npv(flows, rate);
  const { available, grossAnnualReturn } = scenario.capital;

  // The overlay invests only capital unused at time zero. Recurring payments
  // are handled as cash flows in cost NPV, never withdrawn from this balance.
  const investedPrincipal = Math.max(available - timeZeroOutlay, 0);
  const investmentBalance = compound(investedPrincipal, grossAnnualReturn, horizonMonths);
  const foregoneGain =
    timeZeroOutlay * ((1 + grossAnnualReturn) ** (horizonMonths / 12) - 1);
  const investmentAdjustedNpv =
    costNpv + foregoneGain * discountFactor(rate, horizonMonths);

  return {
    path,
    horizonMonths,
    comparable: true,
    reason: null,
    affordable: available + 1e-9 >= timeZeroOutlay,
    timeZeroOutlay,
    costNpv,
    monthlyOwnershipCost: levelMonthlyCost(costNpv, rate, horizonMonths),
    vehicleEquity,
    investedPrincipal,
    investmentBalance,
    investmentAdjustedNpv,
    investmentAdjustedMonthlyCost: levelMonthlyCost(
      investmentAdjustedNpv,
      rate,
      horizonMonths,
    ),
    flows,
  };
}

/**
 * @param {Scenario} scenario
 * @param {number} horizonMonths
 */
export function cashPath(scenario, horizonMonths) {
  const resale = marketValue(scenario.vehicle, horizonMonths);
  const flows = [
    { month: 0, amount: scenario.outTheDoor },
    ...annualCostFlows(annualOwnershipTotal(scenario.ownership), horizonMonths),
    { month: horizonMonths, amount: -resale },
  ];
  return finalize('cash', scenario, horizonMonths, flows, scenario.outTheDoor, resale);
}

/**
 * @param {Scenario} scenario
 * @param {number} horizonMonths
 */
export function loanPath(scenario, horizonMonths) {
  const loan = scenario.loan;
  if (!loan) return notComparable('loan', horizonMonths, 'no loan terms supplied');

  const principal = scenario.outTheDoor - loan.downPayment;
  if (principal < 0) {
    return notComparable('loan', horizonMonths, 'down payment exceeds out-the-door price');
  }

  const payment = loanPayment(principal, loan.apr, loan.termMonths);
  const paidMonths = Math.min(horizonMonths, loan.termMonths);
  const balance = loanBalance(principal, loan.apr, loan.termMonths, paidMonths);
  const resale = marketValue(scenario.vehicle, horizonMonths);

  const flows = [{ month: 0, amount: loan.downPayment }];
  for (let month = 1; month <= paidMonths; month += 1) {
    flows.push({ month, amount: payment });
  }
  flows.push(...annualCostFlows(annualOwnershipTotal(scenario.ownership), horizonMonths));
  flows.push({ month: horizonMonths, amount: -(resale - balance) });

  const result = finalize(
    'loan',
    scenario,
    horizonMonths,
    flows,
    loan.downPayment,
    resale - balance,
  );
  result.monthlyPayment = payment;
  result.outstandingBalance = balance;
  return result;
}

/**
 * @param {Scenario} scenario
 * @param {number} horizonMonths
 * @param {'return'|'buyout'} leaseEnd
 */
export function leasePath(scenario, horizonMonths, leaseEnd = RETURN) {
  const lease = scenario.lease;
  if (!lease) return notComparable('lease', horizonMonths, 'no lease terms supplied');

  const label = horizonMonths <= lease.termMonths ? 'lease' : 'lease-then-buyout';

  if (horizonMonths < lease.termMonths) {
    return notComparable(
      label,
      horizonMonths,
      'holding period ends before lease maturity and no early-termination or buyout quote was supplied',
    );
  }
  if (horizonMonths > lease.termMonths && leaseEnd !== BUYOUT) {
    return notComparable(
      label,
      horizonMonths,
      'holding period exceeds the lease term, which requires a buyout or replacement-lease path',
    );
  }

  const scheduled = lease.firstPaymentInDueAtSigning === false
    ? lease.termMonths
    : lease.termMonths - 1;

  const flows = [{ month: 0, amount: lease.dueAtSigning }];
  for (let month = 1; month <= scheduled; month += 1) {
    flows.push({ month, amount: lease.monthlyPayment });
  }
  flows.push(...annualCostFlows(annualOwnershipTotal(scenario.ownership), horizonMonths));

  let vehicleEquity = 0;
  if (leaseEnd === RETURN) {
    flows.push({ month: lease.termMonths, amount: lease.dispositionFee });
  } else {
    flows.push({
      month: lease.termMonths,
      amount: lease.buyoutPrice + (lease.buyoutFees ?? 0),
    });
    vehicleEquity = marketValue(scenario.vehicle, horizonMonths);
    flows.push({ month: horizonMonths, amount: -vehicleEquity });
  }

  const path = leaseEnd === RETURN ? label : `${label}-buyout`;
  return finalize(path, scenario, horizonMonths, flows, lease.dueAtSigning, vehicleEquity);
}

/**
 * Evaluate every enabled path at one horizon.
 * @param {Scenario} scenario
 * @param {number} horizonMonths
 */
export function comparePaths(scenario, horizonMonths) {
  return [
    cashPath(scenario, horizonMonths),
    loanPath(scenario, horizonMonths),
    leasePath(scenario, horizonMonths, RETURN),
    leasePath(scenario, horizonMonths, BUYOUT),
  ];
}

File: validation/new_car.py
"""Flow 2A (new car acquisition) reference model.

This is a validation harness for the calculation rules in docs/multi-flow-ux.md,
not the production application. It exists to make the documented numbers
reproducible and to catch regressions when the model changes.
"""

from dataclasses import dataclass, field
from typing import List, Optional

from finance import (
    CashFlow,
    annual_cost_flows,
    compound,
    discount_factor,
    exponential_value,
    level_monthly_cost,
    loan_balance,
    loan_payment,
    monthly_rate,
    npv,
)

RETURN = "return"
BUYOUT = "buyout"


@dataclass(frozen=True)
class Vehicle:
    msrp: float
    resale_anchor_month: int
    resale_anchor_value: float

    def market_value(self, month: int) -> float:
        return exponential_value(
            self.msrp, self.resale_anchor_month, self.resale_anchor_value, month
        )


@dataclass(frozen=True)
class Ownership:
    annual_government_charges: float = 0.0
    annual_energy: float = 0.0

    @property
    def annual_total(self) -> float:
        return self.annual_government_charges + self.annual_energy


@dataclass(frozen=True)
class LoanTerms:
    down_payment: float
    apr: float
    term_months: int


@dataclass(frozen=True)
class LeaseTerms:
    due_at_signing: float
    monthly_payment: float
    term_months: int
    disposition_fee: float
    buyout_price: float
    buyout_fees: float = 0.0
    first_payment_in_due_at_signing: bool = True


@dataclass(frozen=True)
class Capital:
    available: float
    gross_annual_return: float


@dataclass(frozen=True)
class Scenario:
    vehicle: Vehicle
    out_the_door: float
    ownership: Ownership
    capital: Capital
    discount_rate_annual: float
    loan: Optional[LoanTerms] = None
    lease: Optional[LeaseTerms] = None


@dataclass
class PathResult:
    path: str
    horizon_months: int
    comparable: bool = True
    reason: Optional[str] = None
    affordable: bool = True
    time_zero_outlay: float = 0.0
    cost_npv: float = 0.0
    monthly_ownership_cost: float = 0.0
    vehicle_equity: float = 0.0
    invested_principal: float = 0.0
    investment_balance: float = 0.0
    investment_adjusted_npv: float = 0.0
    investment_adjusted_monthly_cost: float = 0.0
    flows: List[CashFlow] = field(default_factory=list)


def _not_comparable(path: str, horizon_months: int, reason: str) -> PathResult:
    return PathResult(path=path, horizon_months=horizon_months, comparable=False, reason=reason)


def _finalize(
    result: PathResult,
    scenario: Scenario,
    flows: List[CashFlow],
    time_zero_outlay: float,
    vehicle_equity: float,
) -> PathResult:
    rate = monthly_rate(scenario.discount_rate_annual)
    horizon = result.horizon_months

    result.flows = flows
    result.time_zero_outlay = time_zero_outlay
    result.cost_npv = npv(flows, rate)
    result.monthly_ownership_cost = level_monthly_cost(result.cost_npv, rate, horizon)
    result.vehicle_equity = vehicle_equity

    # The overlay invests only capital left unused at time zero. Recurring
    # payments are handled as cash flows in cost NPV, never withdrawn here.
    result.invested_principal = max(scenario.capital.available - time_zero_outlay, 0.0)
    result.investment_balance = compound(
        result.invested_principal, scenario.capital.gross_annual_return, horizon
    )

    foregone_gain = time_zero_outlay * (
        (1.0 + scenario.capital.gross_annual_return) ** (horizon / 12.0) - 1.0
    )
    result.investment_adjusted_npv = result.cost_npv + foregone_gain * discount_factor(rate, horizon)
    result.investment_adjusted_monthly_cost = level_monthly_cost(
        result.investment_adjusted_npv, rate, horizon
    )
    result.affordable = scenario.capital.available + 1e-9 >= time_zero_outlay
    return result


def cash_path(scenario: Scenario, horizon_months: int) -> PathResult:
    result = PathResult(path="cash", horizon_months=horizon_months)
    resale = scenario.vehicle.market_value(horizon_months)

    flows: List[CashFlow] = [(0, scenario.out_the_door)]
    flows.extend(annual_cost_flows(scenario.ownership.annual_total, horizon_months))
    flows.append((horizon_months, -resale))

    return _finalize(result, scenario, flows, scenario.out_the_door, resale)


def loan_path(scenario: Scenario, horizon_months: int) -> PathResult:
    if scenario.loan is None:
        return _not_comparable("loan", horizon_months, "no loan terms supplied")

    loan = scenario.loan
    result = PathResult(path="loan", horizon_months=horizon_months)
    principal = scenario.out_the_door - loan.down_payment
    if principal < 0:
        return _not_comparable("loan", horizon_months, "down payment exceeds out-the-door price")

    payment = loan_payment(principal, loan.apr, loan.term_months)
    paid_months = min(horizon_months, loan.term_months)
    balance = loan_balance(principal, loan.apr, loan.term_months, paid_months)
    resale = scenario.vehicle.market_value(horizon_months)

    flows: List[CashFlow] = [(0, loan.down_payment)]
    flows.extend((month, payment) for month in range(1, paid_months + 1))
    flows.extend(annual_cost_flows(scenario.ownership.annual_total, horizon_months))
    flows.append((horizon_months, -(resale - balance)))

    return _finalize(result, scenario, flows, loan.down_payment, resale - balance)


def lease_path(scenario: Scenario, horizon_months: int, lease_end: str = RETURN) -> PathResult:
    if scenario.lease is None:
        return _not_comparable("lease", horizon_months, "no lease terms supplied")

    lease = scenario.lease
    label = "lease" if horizon_months <= lease.term_months else "lease-then-buyout"

    if horizon_months < lease.term_months:
        return _not_comparable(
            label,
            horizon_months,
            "holding period ends before lease maturity and no early-termination "
            "or buyout quote was supplied",
        )
    if horizon_months > lease.term_months and lease_end != BUYOUT:
        return _not_comparable(
            label,
            horizon_months,
            "holding period exceeds the lease term, which requires a buyout or "
            "replacement-lease path",
        )

    result = PathResult(path=label if lease_end == RETURN else label + "-buyout",
                        horizon_months=horizon_months)

    scheduled = lease.term_months - 1 if lease.first_payment_in_due_at_signing else lease.term_months
    flows: List[CashFlow] = [(0, lease.due_at_signing)]
    flows.extend((month, lease.monthly_payment) for month in range(1, scheduled + 1))
    flows.extend(annual_cost_flows(scenario.ownership.annual_total, horizon_months))

    if lease_end == RETURN:
        flows.append((lease.term_months, lease.disposition_fee))
        vehicle_equity = 0.0
    else:
        flows.append((lease.term_months, lease.buyout_price + lease.buyout_fees))
        vehicle_equity = scenario.vehicle.market_value(horizon_months)
        flows.append((horizon_months, -vehicle_equity))

    return _finalize(result, scenario, flows, lease.due_at_signing, vehicle_equity)
